src/main/java/org/opensearch/searchrelevance/experiment/HybridOptimizerExperimentProcessor.java

```diff
@@ -143,9 +143,7 @@ public void processHybridOptimizerExperiment(
                 listener
             );
         }).exceptionally(e -> {
-            if (hasFailure.compareAndSet(false, true)) {
```

Comment:
I don't think you should remove this pattern. This approach ensures that failure notification is a single instance. Think of scenario when multiple threads are calling `listener.onFailure()` at the same time, what is the behavior in such case?  
The real issue is in the error propagation chain between the processor and `ExperimentRunningManager`